File: src/main/java/org/cubexmc/model/RedeemIngredient.java
package org.cubexmc.model;

import java.util.Objects;

/**
 * One gem ingredient in a redeem recipe.
 */
public final class RedeemIngredient {
    private final String gemKey;
    private final int amount;

    public RedeemIngredient(String gemKey, int amount) {
        this.gemKey = gemKey == null ? "" : gemKey.trim();
        this.amount = Math.max(1, amount);
    }

    public String getGemKey() {
        return gemKey;
    }

    public int getAmount() {
        return amount;
    }

    public boolean isValid() {
        return !gemKey.isEmpty();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (!(o instanceof RedeemIngredient that))
            return false;
        return amount == that.amount && Objects.equals(gemKey, that.gemKey);
    }

    @Override
    public int hashCode() {
        return Objects.hash(gemKey, amount);
    }

    @Override
    public String toString() {
        return "RedeemIngredient{gemKey='" + gemKey + "', amount=" + amount + '}';
    }
}

File: src/main/java/org/cubexmc/model/RedeemIngredient.kt
package org.cubexmc.model

import java.util.Objects
import kotlin.math.max

/**
 * One gem ingredient in a redeem recipe.
 */
class RedeemIngredient(gemKey: String?, amount: Int) {
    val gemKey: String = gemKey?.trim() ?: ""
    val amount: Int = max(1, amount)

    fun isValid(): Boolean = gemKey.isNotEmpty()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RedeemIngredient) return false
        return amount == other.amount && gemKey == other.gemKey
    }

    override fun hashCode(): Int = Objects.hash(gemKey, amount)

    override fun toString(): String = "RedeemIngredient{gemKey='$gemKey', amount=$amount}"
}
